impl/src/main/java/org/pentaho/di/trans/dataservice/streaming/execution/StreamingServiceTransExecutor.java

```diff
@@ -0,0 +1,241 @@
+/*! ******************************************************************************
+ *
+ * Pentaho Data Integration
+ *
+ * Copyright (C) 2018 by Hitachi Vantara : http://www.pentaho.com
+ *
+ *******************************************************************************
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+package org.pentaho.di.trans.dataservice.streaming.execution;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Throwables;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.RemovalListener;
+import com.google.common.cache.RemovalNotification;
+import io.reactivex.Observable;
+import org.pentaho.di.core.RowMetaAndData;
+import org.pentaho.di.core.exception.KettleException;
+import org.pentaho.di.core.exception.KettleStepException;
+import org.pentaho.di.core.exception.KettleValueException;
+import org.pentaho.di.core.logging.LogChannelInterface;
+import org.pentaho.di.core.row.RowMetaInterface;
+import org.pentaho.di.trans.Trans;
+import org.pentaho.di.trans.dataservice.streaming.StreamList;
+import org.pentaho.di.trans.dataservice.utils.DataServiceConstants;
+import org.pentaho.di.trans.step.RowAdapter;
+import org.pentaho.di.trans.step.StepInterface;
+
+import java.util.Date;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * This class represents a streaming execution for a service transformation.
+ * It spans a thread to run the transformation when data is requested and running thread exists, caching the requests.
+ * When all the cached requests are expired the transformation is stopped and it's associated thread terminated.
+ */
+public class StreamingServiceTransExecutor {
+  private final Trans serviceTrans;
+  private final String id;
+  private final String serviceStepName;
+  private final AtomicBoolean isRunning = new AtomicBoolean( false );
+
+  private StreamList<RowMetaAndData> stepStream;
+
+  /**
+   * Service listener cache.
+   */
+  private final Cache<String, StreamExecutionListener> serviceListeners = CacheBuilder.newBuilder()
+    .concurrencyLevel( 4 )
```

Comment:
this is already the default value, so we can remove it or put a more meaningful value... this number has a performance impact, and should be equal to maximum number of concurrent threads that can access the cache at any given point in time, without incurring in contention. See [javadoc](https://google.github.io/guava/releases/16.0/api/docs/com/google/common/cache/CacheBuilder.html#concurrencyLevel(int)).

Maybe 4 could be a good value (?) but we should put that as a private constant in the class,

Reply:
To be removed. Stayed there because of testing.